test/mock/kernel/mock_framework.zig

```diff
@@ -0,0 +1,588 @@
+                            };
+
+                            break :ret r;
+                        },
+                    };
+
+                    // Re-assign the action list as this would have changed
+                    kv_actions_list.value = action_list;
+                    return ret;
+                } else {
+                    warn("No action list elements for function: " ++ fun_name ++ "\n");
+                    expect(false);
+                    unreachable;
+                }
+            } else {
+                warn("No function name: " ++ fun_name ++ "\n");
+                expect(false);
+                unreachable;
+            }
+        }
+
+        ///
+        /// Initialise the mocking framework.
+        ///
+        /// Return: Self
+        ///     An initialised mocking framework.
+        ///
+        pub fn init() Self {
+            return Self{
+                .named_actions = StringHashMap(ActionList).init(GlobalAllocator),
+            };
+        }
+
+        ///
+        /// End the mocking session. This will check all test parameters and consume functions are
+        /// consumed. Any repeat functions are deinit.
+        ///
+        /// Arguments:
+        ///     IN/OUT self: *Self - Self. This is the mocking object to be modified to finished
+        ///                          the mocking session.
+        ///
+        pub fn finish(self: *Self) void {
+            // Make sure the expected list is empty
+            var it = self.named_actions.iterator();
+            while (it.next()) |next| {
+                var action_list = next.value;
+                if (action_list.popFirst()) |action_node| {
+                    const action = action_node.data;
+                    switch (action.action) {
+                        ActionType.TestValue, ActionType.ConsumeFunctionCall => {
+                            // These need to be all consumed
+                            warn("Unused testing value: Type: {}, value: {} for function '{}'\n", action.action, DataElementType(action.data), next.key);
+                            expect(false);
+                            unreachable;
+                        },
+                        ActionType.RepeatFunctionCall => {
+                            // As this is a repeat action, the function will still be here
+                            // So need to free it
+                            action_list.destroyNode(action_node, GlobalAllocator);
+                            next.value = action_list;
+                        },
+                    }
+                }
+            }
+
+            // Free the function mapping
+            self.named_actions.deinit();
+        }
+    };
+}
+
+/// The global mocking object that is used for a mocking session. Maybe in the future, we can have
+/// local mocking objects so can run the tests in parallel.
+var mock: ?Mock() = null;
```

Comment:
Do we need a global mock and `getMockObject` function, or can the mock be declared locally somewhere?

Reply:
I originally had it locally, but there was a lot of copy paste code that was the same for every file that used the mocking framework, to I put it all in the mocking framework. Plus this ensures, that each mocking session is init and free'ed once for each test. As the comment suggests, if we can get tests to run in parallel, then I can re-think and have local mocking objects.

Reply:
:+1: